server/task/task-memory/src/main/java/org/apache/james/task/SerialTaskManagerWorker.java

```diff
@@ -40,50 +40,66 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import com.google.common.collect.Maps;
 import com.google.common.collect.Sets;
 
 import reactor.core.Disposable;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 import reactor.core.scheduler.Scheduler;
 import reactor.core.scheduler.Schedulers;
-import reactor.util.function.Tuple2;
-import reactor.util.function.Tuples;
 
 public class SerialTaskManagerWorker implements TaskManagerWorker {
     private static final Logger LOGGER = LoggerFactory.getLogger(SerialTaskManagerWorker.class);
     public static final boolean MAY_INTERRUPT_IF_RUNNING = true;
 
     private final Scheduler taskExecutor;
+    private final Scheduler asyncTaskExecutor;
     private final Listener listener;
-    private final AtomicReference<Tuple2<TaskId, CompletableFuture>> runningTask;
+    private final Map<TaskId, CompletableFuture> runningTasks;
     private final Set<TaskId> cancelledTasks;
     private final Duration pollingInterval;
 
     public SerialTaskManagerWorker(Listener listener, Duration pollingInterval) {
         this.pollingInterval = pollingInterval;
         this.taskExecutor = Schedulers.fromExecutor(
             Executors.newSingleThreadExecutor(NamedThreadFactory.withName("task executor")));
+        this.asyncTaskExecutor = Schedulers.fromExecutor(
+            Executors.newCachedThreadPool(NamedThreadFactory.withName("async task executor")));
```

Comment:
I see your point, however I would prefer to keep it open. Once a limited pool is exhausted, it would block any further tasks though they are _supposed_ to run without blocking. And as always, what should be a sensible default limit?
I do not expect this to be a problem in practice, since I assume async safe tasks to be quite rare, compared to regular tasks (Right now, there is just the one ExpireMailboxTask). And an admin starting too many async tasks should not complain about them needing a lot of resources.
If it turns out to be a problem, we can still add a limit through a configuration property later.

Reply:
> I assume async safe tasks to be quite rare, compared to regular tasks (Right now, there is just the one ExpireMailboxTask)

I rather expect that most of our tasks can safely be executed in parallel.

> If it turns out to be a problem, we can still add a limit through a configuration property later.

Ok

Reply:
Do you have any particular long-running task in mind? The tricky part will be to figure out if it is "safe" in the sense of AsyncSafeTask. Might be worth a followup ticket.

Reply:
In my opinion much tasks do not fundamentally alter the state of data, in a way that can create incompatibilities, from a correctness perspective.

EG redelivering dead letter can be run in parallel with mail repository reprocessing.

It then is more to me a resource usage problem: maybe reindexing all of my ElasticSearch data at the same time than my deduplication is not a good idea?

How would we prevent 2 expensive tasks t be executed at the same time?

> Might be worth a followup ticket.

+1

Reply:
> How would we prevent 2 expensive tasks t be executed at the same time?
I would trust a server admin to be sensible enough and not start multiple expensive tasks during peak business hours :)